### DISCUSSION TITLE
Making anylogic and python communicate in an optimzation problem

### DISCUSSION THREAD

**Discussion by TellSol** [the-anylogic-company/AnyLogic-Pypeline#27] [Q&A] [answered]
Hello, 

I'm working on a rather simple optimization problem (Linear Programming), but I am having some issues when it comes to communication between Anylogic and the Python file (which runs the optimization process). 

**1. I have this python script (in the same directory as the .alp file) :**

```

import pyomo.environ as pyo
from pyomo.environ import *
from pyomo.opt import SolverFactory

# An ordinary 2-variable LP optimization problem
#-----------------------------------------------

# PROBLEM:
# We have products A and B, with profits of 350 and 450. Using the constraints
# model.C1 to model.C3, what is the optimal production mix of A and B?


# initialize a concrete model
model = pyo.ConcreteModel()

# define the decision variables
model.a = pyo.Var(bounds=(0,None), within=Integers)
model.b = pyo.Var(bounds=(0,None), within=Integers)

a = model.a
b = model.b

# constraints
model.C1 = pyo.Constraint(expr= 1.5*a + 2*b <= 1000)
model.C2 = pyo.Constraint(expr= 3*a + 4.5*b <= 2000)
model.C3 = pyo.Constraint(expr= 2.5*a + 1.5*b <= 1500)

model.obj = pyo.Objective(expr= 350*a + 450*b, sense=maximize)

# run solver and optimize
opt = SolverFactory('glpk')
opt.solve(model)

# display the optimal solution
a_value = pyo.value(a)
b_value = pyo.value(b)

print("optimal production of product a: ", a_value)
print("optimal production of product b: ", b_value)
```

**2. I have this anylogic model**

![pypelinePic](https://user-images.githubusercontent.com/52788090/159794798-29baa512-224b-47fc-b7d2-044846634ce5.PNG)

So, what I want to achieve is the following:

1. Find a way to pass the values of the parameters "costA" and "costB" to the python script and make them override the values 350 and 450 for model.obj. The data type here is double.

2. Get the variables "a_value" and "b_value" back to the anylogic model, and programatically display them in "text2" and "text3"

3. Make sure that 1. and 2. are triggered by clicking the "Optimize" button. 

Hope it was understandable. Any help is much appreciated! 
Let me know if you need any additional information about the problem, and I will provide it.

**Comment by t-wolfeadam** [comment, marked as answer]
Hi, sorry for such a delay, I'm not sure how I missed this. 

The easiest way to do this is simply to wrap the Python logic in a function that takes the desired arguments and returns the values as a _list_ (this is important). You can then import the function in the model's "On startup" and call it when you click the button.

For example:
```python
# lp_cost_solver.py

# <imports>

# <any other global code>

def get_optimal_values(acost, bcost):
    # ...
    model.obj = pyo.Objective(expr= acost*a + bcost*b, sense=maximize)
    # ...
    return [a_value, b_value] # important to return as a list, not a tuple
```

Inside your "On startup":
```java
pyCommunicator.run("from lp_cost_solver import get_optimal_values");
```

And inside your button:
```java
// Can convert the Python list to a primitive list or an ArrayList (the following does the former)
double[] optValues = pyCommunicator.runResults(double[].class, "get_optimal_values(%s, %s)", costA, costB);
// ... desired reaction, using optValues[0] and optValues[1] ...
```

**Reply by t-wolfeadam** [reply]
If you want to see an alternative method of getting results from an LP solver, see the "Machine Optimizer" example.

But keep in mind that I consider it outdated and not a recommended way to do it, as (1) it uses the old way of converting Python outputs to Java types (via the `getFeedback` function), and (2) it has all the Python code inside the model (specifically in the `solveit` function + Output objects) which makes it harder to maintain and test.

**Reply by TellSol** [reply]
Got it. It was the "double[ ].class" in the button action field i missed. 

Thanks for the clear explanations!

**Reply by TellSol** [reply]
I ran into some trouble again here. The piece of code within the button:

`double[] optValues = pyCommunicator.runResults(double[].class, "get_optimal_values(%s, %s)", costA, costB);`

gave me the error message

`The method runResults(Class<T>, String) in the type PyCommunicator is not applicable for the arguments (Class<double[]>, String, double, double)`


Do you have any ideas to what the issue might be? :)

Any help is much appreciated!

**Reply by t-wolfeadam** [reply]
I'm able to compile using your provided code (using dummy values for the cost) -- 
![image](https://user-images.githubusercontent.com/40703071/170133906-b8425e95-449a-4150-827a-f0ff6de7b5af.png)

So I would make sure you're using the latest version (1.7) or otherwise try restarting AnyLogic, as it's not detecting the string formatting option

**Reply by TellSol** [reply]
Thanks a lot. 

(I replaced costA and costB with the names profitA and profitB as we typically don't want to maximize costs haha:D)


I didn't have the best of luck when using type "double" on my two parameters, so I changed them to type "string", and now I am able to:

1. Build the model without errors
2. Get the expected output in the console (the values 554 and 75)

However, just after receiving the optimal values, I get a whole lot of error messages that says that I have invalid line numbers:

![image](https://user-images.githubusercontent.com/52788090/170583764-537f493a-1f85-425a-a7d5-7e7336dfa325.png)


Any ideas on how to resolve these errors? Are there any changes you suggest for the source attachment configuration?

![image](https://user-images.githubusercontent.com/52788090/170584232-7cd95baa-0ef0-43c0-b618-6c46d3cd74d0.png)

**Reply by t-wolfeadam** [reply]
> I didn't have the best of luck when using type "double" on my two parameters, so I changed them to type "string", and now I am able to

You shouldn't have to do that. The library will convert your double variables to strings for you - it replaces your "get_optimal_values(%s, %s)" with the formatted string ([see here](https://github.com/t-wolfeadam/AnyLogic-Pypeline/blob/master/Source/AL-Pypeline%20Library/Pypeline%20Library.alp#L2625-L2626)).

> I get a whole lot of error messages that says that I have invalid line numbers

That whole block of "at" messages is the [stacktrace](https://stackoverflow.com/a/3988794). You can ignore the line numbers, they're references in the source code and are only relevant to you when it's referring to agents (i.e., Classes) from your model (as you don't have the source of AnyLogic or Pypeline on your system). 

By the error message itself, it looks like you tried to run the function with the "%s" placeholders without providing any arguments. The "invalid syntax" is likely from Python, which doesn't understand what "%s" is. 

While I have custom handling to report Python-specific errors, the references to the Pypeline source code don't match up with the latest version. To be sure, please check the version is 1.7.0 from the [Libraries dialog](https://anylogic.help/advanced/libraries/managing-libraries.html#managing-libraries). To be _really_ sure, take note of the jar location in that dialog, close AnyLogic, and replace the jar file there with a redownloaded one.

**Reply by TellSol** [reply]
Hi again, 

I solved it now, and your proposed solution was correct. The problem was that the model was using the 1.4 Pypeline .jar file, I never actually did the switch from version 1.4 to 1.7. Silly mistake..

Thanks for the patience!
